### PR TITLE
Add Classifier free guidance

### LINKED ISSUE #5825: [RFC]: Classifier-Free Guidance
### Motivation.

I am one of the authors of the paper Stay On Topic with Classifier-Free Guidance ( https://openreview.net/forum?id=RiM3cl9MdK&noteId=s1BXLL1YZD ) who has been nominated as ICML'24 Spotlight Paper. CFG is a sampling technique that allows LLMs to follow the prompt more closely at the cost of two forward passes per token as well as 2 kv caches. CFG brings non trivial improvements overall over standard benchmarks.

I would be extremely interested in having CFG implemented into vLLM. If possible, I would like to get a bit of guidance into the vLLM code base.

### Proposed Change.

CFG contrasts the next token logits between two different prompt (a "positive prompt" a, and a "negative prompt" or "unconditional" b)

Here is the pseudo algorithm

```
while we sample:
    logits_a = log_softmax(model(prompt_a))
    logits_b = log_softmax(model(prompt_b))
    logits = logits_b + cfg_scale * (logits_a - logits_b)
    next_token = sample_from(logits)
    prompt_a.append(next_token)
    prompt_b.append(next_token)
```

As you can see this needs two concurrent kv-caches for an efficient implementation. I tried looking for how Speculative Decoding was implemented but this was quite complex, more than CFG needs.

### Feedback Period.

_No response_

### CC List.

_No response_

### Any Other Things.

I am willing to implement it myself given enough guidance as this looks like a non trivial thing to implement. I think something similar to / reusing bits of Speculative Decoding might be used but the code is non trivial.

### PR DESCRIPTION
**Pull request by zhaoyinglia**
I want to implement classifier-free guidance in vLLM. There’s an issue mentioned by the paper's author here: https://github.com/vllm-project/vllm/issues/5825.

Thanks to @cadedaniel's suggestion, I have implemented a version. Could you provide some advice?

Here are my modifications:

Firstly, I have split the logic of `ModelRunner.execute_model()` into `model_execute`, `get_logits`, `compute_logits`, and `do_sample`.

Additionally, I have added a `CFGWorker` and `CFGModelRunner` to manage the logic of root_worker and guidance_worker. The core logic is as follows:

```
    logits_a = log_softmax(root_worker(prompt_a))
    logits_b = log_softmax(guidance_worker(prompt_b))
    logits = logits_b + cfg_scale * (logits_a - logits_b)
    next_token = sample_from(logits)
    prompt_a.append(next_token)
    prompt_b.append(next_token)
```

Regarding the interface, there is a example: tests/test_cfg_worker.py

The names of the functions and classes are still under discussion.

I look forward to your reply. Thanks!

FIX #5825

**There is a new implementation in https://github.com/zhaoyinglia/vllm/pull/3, which only calls the `model.forward` function once.**

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by cadedaniel** [comment]
Thanks. I'll take a look on Friday.

**Comment by cadedaniel** [comment]
Thanks @zhaoyinglia for the PR. The overall design looks good to me. Unfortunately I don't have bandwidth to shepherd this PR right now. cc @zhuohan123 and @simon-mo, I can help shepherd this algorithmic contribution but will need someone else to own it.

btw I was talking with @zhuohan123 about an algorithmic layer for vLLM, especially for low-latency use-cases. This PR shows the 2nd use-case at this layer, besides speculative decoding. I feel it is in the best interest of vLLM to prioritize support for this layer, happy to chat more.

**Comment by cadedaniel** [comment]
See https://github.com/vllm-project/vllm/pull/6273#issuecomment-2243654991 for more info

**Comment by cadedaniel** [comment]
cc @GindaChen see this for algo design

**Comment by MSLDCherryPick** [comment]
@zhaoyinglia
Hi yinglia, I wonder if you successfully implement CFG in VLLM？

**Comment by zhaoyinglia** [comment]
> @zhaoyinglia Hi yinglia, I wonder if you successfully implement CFG in VLLM？

yes, we already implemented it in our framework, you can find the function here https://github.com/FlagOpen/FlagScale/tree/main/flagscale/inference/core
